### PR TITLE
Don't Change Source CID After Retry

### LINKED ISSUE #1474: Retry Token Makeup
This issue is related to the new Stream 0 dt PR (#1450).

Is a client allowed to change its connection ID in response to a Retry packet from a server?

The reason I ask is that it is possible to conceive of a server generating a token for a stateless Retry that encodes the client's original CID (among other things) in it for validation purposes. If the client is allowed to change its CID, then that obviously would break. So I feel we should have text either explicitly saying the client may (which I don't see how we could prevent) change its CID, and that a server shouldn't try to use the CID for stateless Retry validation or that a client must not change its CID so the server can use it for validation in stateless Retry.

### PR DESCRIPTION
**Pull request by nibanks**
Resolves #1474.

### CODE REVIEW

**Comment by mikkelfj** [comment]
But doesn't this work against fixing loops in retry? Though that hasn't been worked out yet.

**Comment by nibanks** [comment]
I thought he retry loop issue was related changing **destination** connection ID, not source.

**Comment by MikeBishop** [comment]
Agreed.  I thought there was text to the effect that changing the SCID during the handshake was effectively starting a fresh connection attempt, but I don't see it in the draft.  Maybe still in one of the PRs, or maybe that text got lost at some point.

**Comment by mikkelfj** [comment]
Not all in fresh memory but goes like:
Clients sends original random DCID to server.
Server sends retry back to client with a servers SCID
Client uses retry SCID as its new DCID towards the server.
Client uses whatever SCID it happens to prefer, 0 lenght or otherwise, call this Client-SCID.

If the Client-SCID does not change an new random retries happen, the client might be confused. If the server changes Client-SCID it can distinguish responses. But there is a long dicussion that never got fully resolved. Changing Client-SCID is just one option. Reflecting a separate NONCE per retry is probably either better or also needed.

https://github.com/quicwg/base-drafts/issues/1451

**Comment by martinthomson** [comment]
I think that we need MUST NOT here.  Tokens provided in Retry can be validated strictly, and allowing a client to change its source connection ID would make that harder.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -633,7 +633,11 @@ If the client received a Retry packet from the server and sends an Initial
 packet in response, then it sets the Destination Connection ID to the value from
 the Source Connection ID in the Retry packet. Changing Destination Connection ID
 also results in a change to the keys used to protect the Initial packet. It also
-sets the Token field to the token provided in the Retry.
+sets the Token field to the token provided in the Retry. Additionally, the
+client SHOULD NOT change the Source Connection ID because the server may include
+the connection ID as part of its token validation logic. If the client changes
+the Source Connection ID in this case, it would likely trigger a new Retry to be
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: nit: would likely -> could

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -633,7 +633,11 @@ If the client received a Retry packet from the server and sends an Initial
 packet in response, then it sets the Destination Connection ID to the value from
 the Source Connection ID in the Retry packet. Changing Destination Connection ID
 also results in a change to the keys used to protect the Initial packet. It also
-sets the Token field to the token provided in the Retry.
+sets the Token field to the token provided in the Retry. Additionally, the
+client SHOULD NOT change the Source Connection ID because the server may include
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: MUST NOT

**Comment by kazuho** [comment]
@MikeBishop 
> I thought there was text to the effect that changing the SCID during the handshake was effectively starting a fresh connection attempt, but I don't see it in the draft. 

FWIW, my understanding is that the existing text is (emphasis mine):
> During the handshake, an endpoint might receive multiple packets with the long header, and thus be given multiple opportunities to update the Destination Connection ID it sends. A client MUST only change the value it sends in the Destination Connection ID in response to the first packet of each type it receives from the server (Retry or Initial); **a server MUST set its value based on the Initial packet. Any additional changes are not permitted; if subsequent packets of those types include a different Source Connection ID, they MUST be discarded.** This avoids problems that might arise from stateless processing of multiple Initial packets producing different connection IDs.
> [4.7 Connection ID Encoding](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#connection-id-encoding)

Anyways I agree that this is written from server's viewpoint, and I agree that we should have a "MUST NOT" for the client.

**Comment by martinthomson** [comment]
Thanks Nick.  This will make -13.  I consider it editorial (based on the text @kazuho cites, which should have been generic.)  There are a bunch more changes we need to consider with Retry, but this at least clears one of those up.

**Pull request merged**